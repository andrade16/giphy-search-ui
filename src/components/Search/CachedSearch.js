import { SEARCH_CHAR_THRESHOLD } from "../../utils/constants";

class CachedSearch {
  constructor(searchFunction, resultsHandler) {
    this.searchFunction = searchFunction;
    this.resultsHandler = resultsHandler;

    this.cache = {};
  }

  changeQuery(query) {
    if (query.length < SEARCH_CHAR_THRESHOLD) {
      //Only submit a search if it's more than three characters
      this.resultsHandler([]);
      return;
    }
    if (this.cache[query]) {
      // if search is cached, return that data;
      console.log("query retrieved from cache:", query);
      this.resultsHandler(this.cache[query]);
    } else {
      //otherwise, make an API call for the new search
      this.searchFunction(query).then((results) => {
        this.cache[query] = results;
        console.log("query added to cache:", query);
        this.resultsHandler(results);
      });
    }
  }
}

export default CachedSearch;
