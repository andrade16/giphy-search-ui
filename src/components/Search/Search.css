.app-bar-container {
    flex-grow: 1;
}

.app-bar {
    display: flex;
    justify-content: center;
}

#gif-button {
    margin-right: 16px;
}

#gif-title {
    display: none;
    margin-right: 25%;
}

@media(min-width: 600px) {
    #gif-title {
        display: block;
    }
}

.search-container {
    position: relative;
    border-radius: 4px;
    background-color: rgba(255, 255, 255, 0.15);
    margin-right: 16px;
    margin-left: 0;
    width: auto;
}

.search-container:hover {
    background-color: rgba(255, 255, 255, 0.25);
}

#search-button {
    margin: 8px
}

@media (min-width: 960px) {
    #search-input {
        width: 20ch;
    }
}

#search-input {
    font: inherit;
    color: white;
    width: 100%;
    border: 0;
    height: 1.1876em;
    margin: 0;
    display: block;
    min-width: 0;
    background: none;
    box-sizing: content-box;
    letter-spacing: inherit;
    animation-duration: 10ms;
    transition: width 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
    padding: 8px 8px 8px calc(1em + 32px);
}
