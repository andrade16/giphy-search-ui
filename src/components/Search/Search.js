import React, { useState } from "react";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import IconButton from "@material-ui/core/IconButton";
import GifIcon from "@material-ui/icons/Gif";
import Typography from "@material-ui/core/Typography";
import InputBase from "@material-ui/core/InputBase";
import SearchIcon from "@material-ui/icons/Search";
import Button from "@material-ui/core/Button";
import "./Search.css";

const Search = ({ handleQuery }) => {
  const [query, setQuery] = useState("");

  const onQuerySubmit = (event) => {
    event.preventDefault();
    handleQuery(query);
  };

  const onKeyUp = (event) => {
    event.preventDefault();
    if (event.key === "Enter") {
      handleQuery(query);
    }
  };

  return (
    <div className="app-bar-container">
      <AppBar position="sticky" className="app-bar">
        <Toolbar>
          <IconButton
            edge="start"
            disableRipple={true}
            disableFocusRipple={true}
            disableTouchRipple={true}
            id="gif-button"
            href="https://giphy.com/"
            target="_blank"
            color="inherit"
            aria-label="gif link"
          >
            <GifIcon fontSize="large" />
          </IconButton>
          <Typography id="gif-title" variant="h6" noWrap>
            Giphy Search
          </Typography>

          <div className="search-container">
            <InputBase
              placeholder="Enter a searchâ€¦"
              autoComplete="off"
              value={query}
              autoFocus={true}
              id="search-input"
              inputProps={{ "aria-label": "search" }}
              onChange={(event) => setQuery(event.target.value)}
              onKeyUp={onKeyUp}
            />
          </div>
          <div>
            <Button
              variant="contained"
              type="submit"
              color="secondary"
              id="search-button"
              endIcon={<SearchIcon />}
              onClick={onQuerySubmit}
            >
              Search
            </Button>
          </div>
        </Toolbar>
      </AppBar>
    </div>
  );
};

export default Search;
