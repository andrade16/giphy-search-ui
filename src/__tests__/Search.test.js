import React from 'react';
import renderer from 'react-test-renderer';
import Search from '../components/Search/Search';

describe('Search', () => {
    it('renders Search component correctly', () => {
        const tree = renderer.create(<Search/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
})
