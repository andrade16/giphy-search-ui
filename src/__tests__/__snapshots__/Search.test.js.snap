// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders Search component correctly 1`] = `
<div
  className="app-bar-container"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary app-bar MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <a
        aria-disabled={false}
        aria-label="gif link"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
        href="https://giphy.com/"
        id="gif-button"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        target="_blank"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.5 9H13v6h-1.5zM9 9H6c-.6 0-1 .5-1 1v4c0 .5.4 1 1 1h3c.6 0 1-.5 1-1v-2H8.5v1.5h-2v-3H10V10c0-.5-.4-1-1-1zM19 10.5V9h-4.5v6H16v-2h2v-1.5h-2v-1z"
            />
          </svg>
        </span>
      </a>
      <h6
        className="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap"
        id="gif-title"
      >
        Giphy Search
      </h6>
      <div
        className="search-container"
      >
        <div
          className="MuiInputBase-root"
          onClick={[Function]}
        >
          <input
            aria-label="search"
            autoComplete="off"
            autoFocus={true}
            className="MuiInputBase-input"
            id="search-input"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyUp={[Function]}
            placeholder="Enter a searchâ€¦"
            type="text"
            value=""
          />
        </div>
      </div>
      <div>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
          disabled={false}
          id="search-button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          <span
            className="MuiButton-label"
          >
            Search
            <span
              className="MuiButton-endIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </span>
          </span>
        </button>
      </div>
    </div>
  </header>
</div>
`;
